File: packages/sdk/src/ssrf.ts
/**
 * SSRF guard — reject repo URLs whose host is a private / loopback /
 * link-local IP literal. Mirrors `akua-core::ssrf::validate_host`.
 *
 * Limitation: DNS names resolving to private IPs aren't caught here
 * (DNS rebinding is a network-layer problem). Bypass entirely with
 * `AKUA_ALLOW_PRIVATE_HOSTS=1`.
 */

import { AkuaError } from './errors.js';

export class SsrfError extends AkuaError {
  constructor(host: string) {
    super(
      `refusing to fetch from private-range host \`${host}\`. Set AKUA_ALLOW_PRIVATE_HOSTS=1 for local dev.`,
    );
    this.name = 'SsrfError';
  }
}

export function validateHost(host: string): void {
  if (allowPrivate()) return;
  const bare = stripPort(host);
  if (isIpLiteral(bare) && isPrivate(bare)) {
    throw new SsrfError(host);
  }
}

/** Pull out the `host:port` from a full URL and validate. */
export function validateUrl(url: string): void {
  try {
    const u = new URL(url);
    validateHost(u.host);
  } catch (err) {
    if (err instanceof SsrfError) throw err;
    // Malformed URL — let the caller's fetch surface the error.
  }
}

function allowPrivate(): boolean {
  const proc = (globalThis as { process?: { env?: Record<string, string | undefined> } }).process;
  const v = proc?.env?.AKUA_ALLOW_PRIVATE_HOSTS;
  return v === '1' || v === 'true' || v === 'yes';
}

function stripPort(host: string): string {
  // IPv6 bracketed literal: `[::1]:8080`
  if (host.startsWith('[')) {
    const end = host.indexOf(']');
    return end > 0 ? host.slice(1, end) : host;
  }
  const colon = host.lastIndexOf(':');
  if (colon < 0) return host;
  const maybePort = host.slice(colon + 1);
  if (/^\d+$/.test(maybePort)) return host.slice(0, colon);
  return host;
}

function isIpLiteral(host: string): boolean {
  return /^\d+\.\d+\.\d+\.\d+$/.test(host) || host.includes(':');
}

function isPrivate(host: string): boolean {
  // IPv4
  const v4 = host.match(/^(\d+)\.(\d+)\.(\d+)\.(\d+)$/);
  if (v4) {
    const [, a, b] = v4.map(Number) as [number, number, number, number, number];
    if (a === 127) return true; // loopback
    if (a === 10) return true; // RFC1918
    if (a === 172 && b >= 16 && b <= 31) return true;
    if (a === 192 && b === 168) return true;
    if (a === 169 && b === 254) return true; // link-local (AWS metadata)
    if (a === 100 && b >= 64 && b <= 127) return true; // CGNAT
    if (a === 0) return true; // unspecified
    if (a === 255 && b === 255) return true; // broadcast
    return false;
  }
  // IPv6 — crude but covers the common cases. `::1`, `fc00::/7`, `fe80::/10`.
  if (host.includes(':')) {
    const norm = host.toLowerCase();
    if (norm === '::1' || norm === '::') return true;
    // fc00::/7 → first byte 0xfc or 0xfd
    if (/^fc[0-9a-f]{0,2}:/.test(norm) || /^fd[0-9a-f]{0,2}:/.test(norm)) return true;
    // fe80::/10 → fe80-febf
    if (/^fe[89ab][0-9a-f]?:/.test(norm)) return true;
  }
  return false;
}

File: crates/akua-core/src/ssrf.rs
//! SSRF guard — reject repo URLs whose host is a private / loopback /
//! link-local IP literal. Blocks the most obvious attack vectors (cloud
//! metadata at `169.254.169.254`, `127.0.0.1`, RFC1918 ranges) without
//! a full DNS resolver.
//!
//! Limitation: a DNS name resolving to a private IP is *not* caught
//! here — mitigate that at the network layer (egress firewall /
//! resolver allowlist). Opt out entirely with
//! `AKUA_ALLOW_PRIVATE_HOSTS=1` for local development.

use std::net::IpAddr;

use crate::fetch::FetchError;

/// Allow-private bypass env var. Unset by default: private IP literals
/// in `repo:` URLs are rejected. Set to `1` / `true` for local dev.
const ALLOW_PRIVATE_ENV: &str = "AKUA_ALLOW_PRIVATE_HOSTS";

/// Validate a repo host. Rejects bare-IP hosts in private ranges
/// unless the bypass env var is set. DNS names pass through.
pub fn validate_host(host: &str) -> Result<(), FetchError> {
    if allow_private() {
        return Ok(());
    }
    let host_only = strip_port(host);
    // Only IP literals reach here — DNS names aren't parsable.
    if let Ok(ip) = host_only.parse::<IpAddr>() {
        if is_private(ip) {
            return Err(FetchError::PrivateHost {
                host: host.to_string(),
            });
        }
    }
    Ok(())
}

fn allow_private() -> bool {
    matches!(
        std::env::var(ALLOW_PRIVATE_ENV).ok().as_deref(),
        Some("1") | Some("true") | Some("yes")
    )
}

/// Strip an optional `:port` (IPv6 literals get bracket-stripped too).
fn strip_port(host: &str) -> &str {
    if let Some(rest) = host.strip_prefix('[') {
        if let Some(end) = rest.find(']') {
            return &rest[..end];
        }
    }
    match host.rsplit_once(':') {
        Some((h, port)) if port.chars().all(|c| c.is_ascii_digit()) => h,
        _ => host,
    }
}

fn is_private(ip: IpAddr) -> bool {
    match ip {
        IpAddr::V4(v4) => {
            v4.is_loopback()       // 127/8
                || v4.is_private() // 10/8, 172.16/12, 192.168/16
                || v4.is_link_local() // 169.254/16 (includes AWS metadata)
                || v4.is_broadcast()
                || v4.is_unspecified() // 0.0.0.0
                || v4.octets()[0] == 100 && (v4.octets()[1] & 0xC0) == 0x40 // CGNAT 100.64/10
        }
        IpAddr::V6(v6) => {
            v6.is_loopback()
                || v6.is_unspecified()
                || (v6.segments()[0] & 0xfe00) == 0xfc00  // fc00::/7 ULA
                || (v6.segments()[0] & 0xffc0) == 0xfe80  // fe80::/10 link-local
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_aws_metadata_ip() {
        assert!(validate_host("169.254.169.254").is_err());
    }

    #[test]
    fn rejects_loopback() {
        assert!(validate_host("127.0.0.1").is_err());
        assert!(validate_host("127.0.0.1:8080").is_err());
        assert!(validate_host("[::1]").is_err());
    }

    #[test]
    fn rejects_rfc1918() {
        assert!(validate_host("10.0.0.1").is_err());
        assert!(validate_host("172.16.5.5").is_err());
        assert!(validate_host("192.168.1.1").is_err());
    }

    #[test]
    fn allows_public_ip() {
        assert!(validate_host("8.8.8.8").is_ok());
        assert!(validate_host("1.1.1.1:443").is_ok());
    }

    #[test]
    fn allows_dns_names() {
        // DNS names aren't IP-parseable — we trust them (consumer's
        // network layer handles DNS-rebinding mitigations).
        assert!(validate_host("ghcr.io").is_ok());
        assert!(validate_host("charts.bitnami.com").is_ok());
        assert!(validate_host("registry.example.com:5000").is_ok());
    }

    #[test]
    fn env_var_bypass() {
        static LOCK: std::sync::Mutex<()> = std::sync::Mutex::new(());
        let _guard = LOCK.lock().unwrap_or_else(|e| e.into_inner());
        let _g = super::super::test_util::ScopedEnvVar::set("AKUA_ALLOW_PRIVATE_HOSTS", "1");
        assert!(validate_host("127.0.0.1").is_ok());
    }
}
